File: android/app/src/test/kotlin/sh/kavi/fasttravel/core/EffectiveIgnoreListTest.kt
package sh.kavi.fasttravel.core

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import sh.kavi.fasttravel.data.AutoIgnoreStore

class EffectiveIgnoreListTest {

    private fun c(count: Int, dni: Boolean = false) =
        AutoIgnoreStore.Candidate(count = count, doNotIgnore = dni)

    @Test
    fun `permanent entries are always included`() {
        assertEquals(
            listOf("cat"),
            effectiveIgnoreList(
                permanent = listOf("cat"),
                local = emptySet(),
                candidates = emptyMap(),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `local entries are always included`() {
        assertEquals(
            listOf("scholr"),
            effectiveIgnoreList(
                permanent = emptyList(),
                local = setOf("scholr"),
                candidates = emptyMap(),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `baseline, local and active candidates merge in order, deduped lowercase`() {
        assertEquals(
            listOf("base", "mine", "auto"),
            effectiveIgnoreList(
                permanent = listOf("BASE"),
                local = setOf("Mine", "base"), // "base" dups the baseline
                candidates = mapOf("auto" to c(count = 3), "low" to c(count = 1)),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `candidate at or above threshold is included`() {
        assertEquals(
            listOf("fcb"),
            effectiveIgnoreList(
                permanent = emptyList(),
                local = emptySet(),
                candidates = mapOf("fcb" to c(count = 3)),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `candidate below threshold is excluded`() {
        assertEquals(
            emptyList<String>(),
            effectiveIgnoreList(
                permanent = emptyList(),
                local = emptySet(),
                candidates = mapOf("fcb" to c(count = 2)),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `do-not-ignore candidate is excluded regardless of count`() {
        assertEquals(
            emptyList<String>(),
            effectiveIgnoreList(
                permanent = emptyList(),
                local = emptySet(),
                candidates = mapOf("fcb" to c(count = 10, dni = true)),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `permanent wins even if also flagged DNI as candidate`() {
        assertEquals(
            listOf("fcb"),
            effectiveIgnoreList(
                permanent = listOf("fcb"),
                local = emptySet(),
                candidates = mapOf("fcb" to c(count = 10, dni = true)),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `local wins even if also flagged DNI as candidate`() {
        assertEquals(
            listOf("fcb"),
            effectiveIgnoreList(
                permanent = emptyList(),
                local = setOf("fcb"),
                candidates = mapOf("fcb" to c(count = 10, dni = true)),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `result is lowercase and deduplicated`() {
        assertEquals(
            listOf("cat"),
            effectiveIgnoreList(
                permanent = listOf("CAT", "cat"),
                local = emptySet(),
                candidates = mapOf("cat" to c(count = 5, dni = false)),
                threshold = 3,
            ),
        )
    }

    @Test
    fun `empty inputs return empty list`() {
        assertEquals(
            emptyList<String>(),
            effectiveIgnoreList(
                permanent = emptyList(),
                local = emptySet(),
                candidates = emptyMap(),
                threshold = 3,
            ),
        )
    }
}

File: extension/tests/unit/effective-ignore-list.test.ts
import { describe, it, expect } from "vitest";
import { effectiveIgnoreList } from "../../src/core/effective-ignore-list.js";

const c = (count: number, doNotIgnore = false) => ({ count, doNotIgnore });

describe("effectiveIgnoreList", () => {
  it("permanent entries are always included", () => {
    expect(effectiveIgnoreList(["cat", "dog"], {}, 3)).toEqual(["cat", "dog"]);
  });

  it("candidate at or above threshold is included (boundary count == threshold)", () => {
    expect(effectiveIgnoreList([], { fcb: c(3) }, 3)).toEqual(["fcb"]);
    expect(effectiveIgnoreList([], { fcb: c(5) }, 3)).toEqual(["fcb"]);
  });

  it("candidate below threshold is excluded", () => {
    expect(effectiveIgnoreList([], { fcb: c(2) }, 3)).toEqual([]);
  });

  it("do-not-ignore candidate is excluded regardless of count", () => {
    expect(effectiveIgnoreList([], { fcb: c(10, true) }, 3)).toEqual([]);
  });

  it("permanent wins even if also flagged DNI as candidate (single entry in output)", () => {
    expect(
      effectiveIgnoreList(["fcb"], { fcb: c(0, true) }, 3),
    ).toEqual(["fcb"]);
  });

  it("result is lowercase and deduplicated", () => {
    expect(
      effectiveIgnoreList(["CAT", "cat"], { cat: c(5) }, 3),
    ).toEqual(["cat"]);
  });

  it("empty inputs return empty array", () => {
    expect(effectiveIgnoreList([], {}, 3)).toEqual([]);
  });
});
